src/storage/storage-facade.ts

```diff
@@ -400,48 +401,85 @@ class DexieStorageProvider implements IStorageInterface {
     return false;
   }
 
-  async syncData() {
+  ensureFirebaseConnection(): Promise<FireStoreStorageProvider> {
     const fsProvider = new FireStoreStorageProvider();
+    if(! this.haveFireStoreConnection) {
+      const portalData = this.portalData;
+      const appName = portalData?.database.appName ?? "report-service-dev";
+      return fsProvider
+        .initializeDB({name: appName, preview: false, offline: false, portalData})
+        .then(() =>  {
+          this.haveFireStoreConnection = true;
+          return fsProvider;
+        })
+        .catch((e) => {
+          console.error(e);
+          console.warn("reinitialization of database?");
+          return Promise.resolve(fsProvider);
+        });
+    }
+    return Promise.resolve(fsProvider);
+  }
+
+  // This is hack to simulate requests taking some time.
+  fakeWaitingPromise(timeToResolve=5000) {
+    return new Promise((res) => window.setTimeout(()=> res(), timeToResolve));
+  }
+
+  syncData(): Promise<boolean> {
     const portalData = this.portalData;
-    const appName = portalData?.database.appName ?? "report-service-dev";
     if(portalData) {
-      try {
-        if(! this.haveFireStoreConnection) {
-          await fsProvider.initializeDB({ name: appName, preview: false, offline: false, portalData});
-          this.haveFireStoreConnection = true;
-        }
-        // FIXME: this should be checking the resourceUrl of the rawPoralData
-        // we don't want to send answers for the wrong resource url to an assignment
-        const answers = await dexieStorage
-          .answers
-          .where({resource_url: _currentOfflineResourceUrl})
-          .toArray();
-        console.dir(answers);
-        for(const answer of answers) {
-          fsProvider.createOrUpdateAnswer(answer);
-        }
-        // TODO: We need to track dirty state, and clear the dirty flag here.
-        // Now do plugin states. No need really, so lets skip it for now.
-        // const pluginStates = await dexieStorage
-        // .pluginStates
-        // .toArray();
-
-        // for(const pState of pluginStates) {
-        //   const {pluginId, state} = pState;
-        //   if(state) {
-        //     fsProvider.setLearnerPluginState(pluginId, state);
-        //   }
-        // }
-      }
+      // DataSyncTracker will emit an event that tells plugins
+      // that we are online, and its time to save data:
+      const mySyncTracker = new DataSyncTracker(60 * 30, 5);
+      const portalResourceUrl = portalData.resourceUrl;
+      // Save student answers ...
+      const answerSavingPromise = this.ensureFirebaseConnection()
+        .then((fsProvider)  => {
+          dexieStorage.answers
+            .where({resource_url: portalResourceUrl})
```

Comment:
This might still result in some weird behavior where I'm looking at an activity in the APO that is different than the portalData.resourceUrl and then doing a sync will cause it to upload data for a different activity.
It might be safer to raise an error if currentOfflineResourceUrl doesn't match the portalResourceUrl.

Reply:
We can disable the button in that case where the current_activity doesn't match the portals resource URL too.

Reply:
The new change looks good.